qa/rolling-upgrade/src/javaRestTest/java/org/elasticsearch/upgrades/SystemIndexMappingUpgradeIT.java

```diff
@@ -0,0 +1,165 @@
+import org.elasticsearch.test.rest.RestTestLegacyFeatures;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.rules.RuleChain;
+import org.junit.rules.TemporaryFolder;
+import org.junit.rules.TestRule;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+
+import static org.elasticsearch.test.LambdaMatchers.transformedMatch;
+import static org.hamcrest.Matchers.allOf;
+import static org.hamcrest.Matchers.anEmptyMap;
+import static org.hamcrest.Matchers.anyOf;
+import static org.hamcrest.Matchers.everyItem;
+import static org.hamcrest.Matchers.hasKey;
+import static org.hamcrest.Matchers.not;
+
+public class SystemIndexMappingUpgradeIT extends ESRestTestCase {
+
+    private static final String OLD_CLUSTER_VERSION = System.getProperty("tests.old_cluster_version");
+    private static final TemporaryFolder repoDirectory = new TemporaryFolder();
+
+    private static final ElasticsearchCluster cluster = ElasticsearchCluster.local()
+        .distribution(DistributionType.DEFAULT)
+        .version(Version.fromString(OLD_CLUSTER_VERSION))
+        .nodes(3)
+        .setting("path.repo", new Supplier<>() {
+            @Override
+            @SuppressForbidden(reason = "TemporaryFolder only has io.File methods, not nio.File")
+            public String get() {
+                return repoDirectory.getRoot().getPath();
+            }
+        })
+        .setting("xpack.security.enabled", "false")
+        .build();
+
+    @ClassRule
+    public static TestRule ruleChain = RuleChain.outerRule(repoDirectory).around(cluster);
+
+    private String testRestCluster = cluster.getHttpAddresses();
+
+    @Before
+    public void setup() {
+        assumeFalse(
+            "Testing upgrades from before CompatibilityVersions held mapping versions in cluster state",
+            testFeatureService.clusterHasFeature(RestTestLegacyFeatures.SYSTEM_INDICES_MAPPING_VERSION_IN_CLUSTER_STATE.id())
+        );
+    }
+
+    @Override
+    protected String getTestRestCluster() {
+        return testRestCluster;
+    }
+
+    @Override
+    protected final Settings restClientSettings() {
+        return Settings.builder()
+            .put(super.restClientSettings())
+            // increase the timeout here to 90 seconds to handle long waits for a green
+            // cluster health. the waits for green need to be longer than a minute to
+            // account for delayed shards
+            .put(ESRestTestCase.CLIENT_SOCKET_TIMEOUT, "90s")
+            .build();
+    }
+
+    private static String getNodesAttribute(ObjectPath objectPath, String nodeId, String attribute) {
+        try {
+            return objectPath.evaluate("nodes." + nodeId + "." + attribute);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static List<String> getUpgradedNodesAddresses() throws IOException {
```

Comment:
There must be a helper method somewhere that does this already?

Reply:
Couldn't find one, but I'll look again